narwhals/translate.py

```diff
@@ -223,11 +236,180 @@ def get_native_namespace(obj: Any) -> Any:
     return obj.__native_namespace__()
 
 
+def narwhalify(
+    func: Callable[..., Any] | None = None,
+    *,
+    strict: bool = True,
+    eager_only: bool | None = None,
+    series_only: bool | None = None,
+    allow_series: bool | None = None,
+) -> Callable[..., Any]:
+    """
+    Decorate a function so that the first argument gets narwhalified with `strict=False`.
+
+    Using
+
+    ```python
+    import narwhals as nw
+
+    @nw.narwhalify
+    def func(df):
+        return df.group_by('a').agg(nw.col('b').sum())
+    ```
+
+    is shorthand for
+
+    ```python
+    import narwhals as nw
+
+    def func(df_any):
+        df = nw.from_native(df_any, strict=False)
+        df = df.group_by('a').agg(nw.col('b').sum())
+        return nw.to_native(df)
+    ```
+
+    Arguments:
+        native_dataframe: Raw dataframe from user.
+            Depending on the other arguments, input object can be:
+
+            - pandas.DataFrame
+            - polars.DataFrame
+            - polars.LazyFrame
+            - anything with a `__narwhals_dataframe__` or `__narwhals_lazyframe__` method
+            - pandas.Series
+            - polars.Series
+            - anything with a `__narwhals_series__` method
+        strict: Whether to raise if object can't be converted (default) or
+            to just leave it as-is.
+        eager_only: Whether to only allow eager objects.
+        series_only: Whether to only allow series.
+        allow_series: Whether to allow series (default is only dataframe / lazyframe).
+    """
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            if args:
+                df_any = args[0]
+            elif kwargs:
+                first_key = next(iter(kwargs.keys()))
+                df_any = kwargs[first_key]
+            else:
+                raise TypeError("Expected function which takes at least one argument.")
```

Comment:
Maybe this detail should be explained in the docstring? i.e. best practice is to wrap a function that takes a dataframe as first argument (or otherwise explicitly pass it as first keyword argument)

Reply:
> otherwise explicitly pass it as first keyword argument

yeah this is the part i'm not keen on...it's unexpected to users if they _have_ to pass arguments in a certain order...will see what we can do about this